src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,285 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal static class SocksHelper
+    {
+        // socks protocol limits address length to 1 byte, thus the maximum possible buffer is 256+other fields
+        private const int BufferSize = 512;
+        private const int ProtocolVersion4 = 4;
+        private const int ProtocolVersion5 = 5;
+        private const byte METHOD_NO_AUTH = 0;
+        // private const byte METHOD_GSSAPI = 1;
+        private const byte METHOD_USERNAME_PASSWORD = 2;
+        private const byte METHOD_NO_ACCEPTABLE = 0xFF;
+        private const byte CMD_CONNECT = 1;
+        // private const byte CMD_BIND = 2;
+        // private const byte CMD_UDP_ASSOCIATE = 3;
+        private const byte ATYP_IPV4 = 1;
+        private const byte ATYP_DOMAIN_NAME = 3;
+        private const byte ATYP_IPV6 = 4;
+        private const byte REP_SUCCESS = 0;
+        // private const byte REP_FAILURE = 1;
+        // private const byte REP_NOT_ALLOWED = 2;
+        // private const byte REP_NETWORK_UNREACHABLE = 3;
+        // private const byte REP_HOST_UNREACHABLE = 4;
+        // private const byte REP_CONNECTION_REFUSED = 5;
+        // private const byte REP_TTL_EXPIRED = 6;
+        // private const byte REP_CMD_NOT_SUPPORT = 7;
+        // private const byte REP_ATYP_NOT_SUPPORT = 8;
+        private const byte CD_SUCCESS = 90;
+
+        public static ValueTask EstablishSocksTunnelAsync(Stream stream, string host, int port, Uri proxyUri, ICredentials? proxyCredentials, CancellationToken cancellationToken)
```

Comment:
We should be flowing the `async` flag into all these methods.

Reply:
Since `sync` reads/writes don't accept the `CancellationToken`, cancellation has to be done by disposing the stream itself.

We also want to dispose the stream in case we're throwing, so one way this could be done is:
- Making this method async
- Registering for cancellation
- Not passing the CT forward